sdk/dotnet/Pulumi/Serialization/Serializer.cs

```diff
@@ -0,0 +1,296 @@
+                var isKnown = data.IsKnown;
+                var isSecret = data.IsSecret;
+
+                if (!isKnown)
+                    return Constants.UnknownValue;
+
+                var value = await SerializeAsync($"{ctx}.id", data.Value).ConfigureAwait(false);
+                if (isSecret)
+                {
+                    var builder = ImmutableDictionary.CreateBuilder<string, object?>();
+                    builder.Add(Constants.SpecialSigKey, Constants.SpecialSecretSig);
+                    builder.Add(Constants.SecretValueName, value);
+                    return builder.ToImmutable();
+                }
+
+                return value;
+            }
+
+            if (prop is CustomResource customResource)
+            {
+                // Resources aren't serializable; instead, we serialize them as references to the ID property.
+                if (_excessiveDebugOutput)
+                {
+                    Log.Debug($"Serialize property[{ctx}]: Encountered CustomResource");
+                }
+
+                this.DependentResources.Add(customResource);
+                return await SerializeAsync($"{ctx}.id", customResource.Id).ConfigureAwait(false);
+            }
+
+            if (prop is ComponentResource componentResource)
+            {
+                // Component resources often can contain cycles in them.  For example, an awsinfra
+                // SecurityGroupRule can point a the awsinfra SecurityGroup, which in turn can point
+                // back to its rules through its 'egressRules' and 'ingressRules' properties.  If
+                // serializing out the 'SecurityGroup' resource ends up trying to serialize out
+                // those properties, a deadlock will happen, due to waiting on the child, which is
+                // waiting on the parent.
+                //
+                // Practically, there is no need to actually serialize out a component.  It doesn't
+                // represent a real resource, nor does it have normal properties that need to be
+                // tracked for differences (since changes to its properties don't represent changes
+                // to resources in the real world).
+                //
+                // So, to avoid these problems, while allowing a flexible and simple programming
+                // model, we just serialize out the component as its urn.  This allows the component
+                // to be identified and tracked in a reasonable manner, while not causing us to
+                // compute or embed information about it that is not needed, and which can lead to
+                // deadlocks.
+                if (_excessiveDebugOutput)
+                {
+                    Log.Debug($"Serialize property[{ctx}]: Encountered ComponentResource");
+                }
+
+                return await SerializeAsync($"{ctx}.urn", componentResource.Urn).ConfigureAwait(false);
+            }
```

Comment:
Is this consistent with what we do in the other runtimes? I don't think we handled `ComponentResource`s specially at all in those cases. Is this necessary strictly to break cycles?

Reply:
this is consistent with what we do with the JS/TS sdk.